samcli/commands/local/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,759 @@
+
+        Parameters
+        ----------
+        intrinsic: dict, str, list, bool, int
+            This is an intrinsic property or an intermediate step
+        parent_function: str
+            In case there is a missing property, this is used to figure out where the property resolved is missing.
+
+        Return
+        ---------
+        The simplified version of the intrinsic function. This could be a list,str,dict depending on the format required
+        """
+        if intrinsic is None:
+            raise InvalidIntrinsicException("Missing Intrinsic property in {}".format(parent_function))
+
+        if any(isinstance(intrinsic, object_type) for object_type in [string_types, list, bool, int]):
+            return intrinsic
+
+        if intrinsic == {}:
+            return intrinsic
+
+        if not isinstance(intrinsic, dict):
+            raise InvalidIntrinsicException(
+                "Invalid Intrinsic type. It is not a int, list, str, or dict {}".format(intrinsic))
+
+        keys = list(intrinsic.keys())
+        key = keys[0]
+
+        if key in self.intrinsic_key_function_map:
+            intrinsic_value = intrinsic.get(key)
+            return self.intrinsic_key_function_map.get(key)(intrinsic_value)
+        elif key in self.conditional_key_function_map:
+            intrinsic_value = intrinsic.get(key)
+            return self.conditional_key_function_map.get(key)(intrinsic_value)
+
+        # In this case, it is a dictionary that doesn't directly contain an intrinsic resolver and must be
+        # re-parsed to resolve.
+        sanitized_dict = {}
+        for key, val in intrinsic.items():
+            sanitized_key = self.intrinsic_property_resolver(key, parent_function=parent_function)
+            sanitized_val = self.intrinsic_property_resolver(val, parent_function=parent_function)
+            verify_intrinsic_type_str(sanitized_key,
+                                      message="The keys of the dictionary {} in {} must all resolve to a string"
+                                      .format(sanitized_key, parent_function))
+            sanitized_dict[sanitized_key] = sanitized_val
+        return sanitized_dict
+
+    def resolve_template(self, ignore_errors=False):
+        """
+        This will parse through every entry in a CloudFormation template and resolve them based on the symbol_resolver.
+        Customers can optionally ignore resource errors and default to whatever the resource provides.
+
+        Return
+        -------
+        A resolved template with all references possible simplified
+        """
+        processed_template = {}
+        for key, val in self.resources.items():
+            processed_key = self.symbol_resolver.get_translation(key, IntrinsicResolver.REF) or key
+            try:
+                processed_resource = self.intrinsic_property_resolver(val)
+                processed_template[processed_key] = processed_resource
```

Comment:
Not sure I follow what is going on here. Isn't `processed_key` always equal to `key` (aka. the logical id)?

Reply:
There are cases when processed_key != key. The example in the SAM Ref code was RestApi.Deployment -> NewRestApi. This will be updated based on parameters to the logical_id_translator.